fdbclient/include/fdbclient/SimKmsVault.cpp

```diff
@@ -0,0 +1,181 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/SimKmsVault.h"
+#include "fdbclient/BlobCipher.h"
+#include "fdbclient/ClientKnobs.h"
+
+#include "fdbclient/Knobs.h"
+#include "flow/Arena.h"
+#include "flow/EncryptUtils.h"
+#include "flow/FastRef.h"
+#include "flow/IRandom.h"
+#include "flow/network.h"
+
+SimKmsVaultKeyCtx::SimKmsVaultKeyCtx(EncryptCipherBaseKeyId kId, const uint8_t* data, const int dataLen)
+  : id(kId), keyLen(dataLen) {
+	key = makeString(keyLen);
+	memcpy(mutateString(key), data, keyLen);
+	kcv = Sha256KCV().computeKCV((const uint8_t*)data, dataLen);
+	if (DEBUG_SIM_KEY_CIPHER) {
+		TraceEvent(SevDebug, "SimKmsVaultKeyCtxInit")
+		    .detail("BaseCipherId", kId)
+		    .detail("BaseCipherLen", dataLen)
+		    .detail("KCV", kcv);
+	}
+}
+
+int SimKmsVaultKeyCtx::getKeyLen(const EncryptCipherBaseKeyId id) {
+	ASSERT_GT(id, INVALID_ENCRYPT_CIPHER_KEY_ID);
+
+	int ret = AES_256_KEY_LENGTH;
+	if ((id % 2) == 0) {
+		ret += (id % (MAX_BASE_CIPHER_LEN - AES_256_KEY_LENGTH));
+	}
+	CODE_PROBE(ret == AES_256_KEY_LENGTH, "SimKmsVault BaseCipherKeyLen AES_256_KEY_LENGTH");
+	CODE_PROBE(ret != AES_256_KEY_LENGTH, "SimKmsVault BaseCipherKeyLen variable length");
+
+	return ret;
+}
+
+class SimKmsVaultCtx : NonCopyable, public ReferenceCounted<SimKmsVaultCtx> {
+public:
+	// Public visibility constructior ONLY to assist FlowSingleton instance creation.
+	// API Note: Constructor is expected to be instantiated only in simulation mode.
+
+	explicit SimKmsVaultCtx(bool ignored) {
+		ASSERT(g_network->isSimulated());
+		init();
+	}
+
+	static Reference<SimKmsVaultCtx> getInstance() {
+		if (g_network->isSimulated()) {
+			return FlowSingleton<SimKmsVaultCtx>::getInstance(
+			    []() { return makeReference<SimKmsVaultCtx>(g_network->isSimulated()); });
+		} else {
+			static SimKmsVaultCtx instance;
+			return Reference<SimKmsVaultCtx>::addRef(&instance);
+		}
+	}
+
+	Reference<SimKmsVaultKeyCtx> getByBaseCipherId(const EncryptCipherBaseKeyId baseCipherId) {
+		auto itr = keyvault.find(baseCipherId);
+		if (itr == keyvault.end()) {
+			return Reference<SimKmsVaultKeyCtx>();
+		}
+		return itr->second;
+	}
+
+	EncryptCipherBaseKeyId getBaseCipherIdFromDomainId(const EncryptCipherDomainId domainId) const {
+		return 1 + abs(domainId) % maxEncryptionKeys;
+	}
+
+	uint32_t maxKeys() const { return maxEncryptionKeys; }
+
+private:
+	SimKmsVaultCtx() { init(); }
+
+	void init() {
+		maxEncryptionKeys = CLIENT_KNOBS->SIM_KMS_VAULT_MAX_KEYS;
+		// Ensure Vault provides 'stable encryption key semantics' across process restarts
+		const unsigned char SHA_KEY[] = "0c39e7906db6d51ac0573d328ce1b6be";
```

Comment:
As mentioned in the comment above, the `SHA_KEY` provides the `stable encryption key` semantics to the vault. 

SimKmsVault uses HMAC_SHA to generate encryption keys on the startup, hence, for determinism we use static `SHA_KEY` and known set of `data` fed to the algorithm on every process start to generate the same set of keys.

```
			computeAuthToken({ { reinterpret_cast<const uint8_t*>(&i), sizeof(i) } },
			                 SHA_KEY,
			                 AES_256_KEY_LENGTH,
			                 &digest[0],
			                 EncryptAuthTokenAlgo::ENCRYPT_HEADER_AUTH_TOKEN_ALGO_HMAC_SHA,
			                 AUTH_TOKEN_HMAC_SHA_SIZE);
```

If needed I can elaborate the existing comment `Ensure Vault provides 'stable encryption key semantics'.
Thanks

Reply:
@sfc-gh-jslocum Updated the code with more elaborate documentation. Hope it helps.

Reply:
I think I mis-phrased my question - I meant how was the specific SHA_KEY "0c39e7906db6d51ac0573d328ce1b6be" chosen as the one to use? Since normally this key is intentionally a secret and not checked in to a repo? I assume it was just randomly generated bytes but i wanted to confirm

Reply:
Ah my bad, I should have addressed it as well. 

Yes, it is just a random key generated out of `thin air`. Given FDB doesn't implement `Vault` functionality, so we don't have logic to generate/rotate and store encryption keys. The simulated version of Vault is an attempt to fill the gap and assist with FDB simulation runs (remove dependency of external KMS)